hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/upgrade/OMLayoutFeatureAspect.java

```diff
@@ -63,16 +68,47 @@ public void checkLayoutFeature(JoinPoint joinPoint) throws Throwable {
         lvm = new OMLayoutVersionManager();
       }
     }
+    checkIsAllowed(joinPoint.getSignature().toShortString(), lvm, featureName);
+  }
+
+  private void checkIsAllowed(String operationName,
+                              LayoutVersionManager lvm,
+                              String featureName) throws OMException {
     if (!lvm.isAllowed(featureName)) {
       LayoutFeature layoutFeature = lvm.getFeature(featureName);
       throw new OMException(String.format("Operation %s cannot be invoked " +
               "before finalization. It belongs to the layout feature %s, " +
               "whose layout version is %d. Current Layout version is %d",
-          joinPoint.getSignature().toShortString(),
+          operationName,
           layoutFeature.name(),
           layoutFeature.layoutVersion(),
           lvm.getMetadataLayoutVersion()),
           NOT_SUPPORTED_OPERATION);
     }
   }
+
+  @Pointcut("execution(* " +
+      "org.apache.hadoop.ozone.om.request.OMClientRequest+.preExecute(..)) " +
+      "&& @this(org.apache.hadoop.ozone.om.upgrade.BelongsToLayoutVersion)")
+  public void omRequestPointCut() {
+  }
+
+  @Before("omRequestPointCut()")
+  public void beforeRequestApplyTxn(final JoinPoint joinPoint)
```

Comment:
Could you please add some example case to show the usage of this preExecute hook. Thanks!

Reply:
To bring in the FSO through upgrades, we would need to

1. Add a new Layout feature in org.apache.hadoop.ozone.om.upgrade.OMLayoutFeature
2. To make sure new feature is not used in finalization, we can do either of the following
    - add the @BelongsToLayoutVersion to all the new FSO request classes referencing the above layout feature
    - Create a prefinalize validation action like org.apache.hadoop.hdds.scm.server.upgrade.ScmHAUnfinalizedStateValidationAction to make sure one cannot enable FSO layout version before finalization.
cc @errose28

Reply:
An example usage --> org.apache.hadoop.ozone.om.upgrade.MockOmRequest. Note that 'preExecute' method is not mandatory for OmRequest class. If not provided, it is used from parent OmClientRequest class.

Reply:
Thanks @avijayanhwx for the example case. IIUC the layout version concept, this allows to build a hierarchy of features or develop new feature on top of an existing layout feature safely.
For example, this is the hierarchy of feature list : @BelongsToLayoutVersion("INITIAL_VERSION") -> @BelongsToLayoutVersion("FSO") -> @BelongsToLayoutVersion("NFS") -> @BelongsToLayoutVersion("POSIX")

am I correct?

```
AbstractLayoutVersionManager.java

  public boolean isAllowed(LayoutFeature layoutFeature) {
    return layoutFeature.layoutVersion() <= metadataLayoutVersion;
  }
```

Reply:
Yes, that is correct. The hierarchy is simple now, where the layout versions are a single line of increasing numbers. In the future, we may support more complex use cases like dependent layout features.

Reply:
@avijayanhwx, ok, makes sense to me!